### ISSUE TITLE
💅 noImplicitAnyLet false positives when assigning in if statement

### ISSUE THREAD

**Issue by LinusU** [biomejs/biome#7386]
### Environment information

```bash
CLI:
  Version:                      2.0.6
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-ghostty
  JS_RUNTIME_VERSION:           v22.14.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/10.9.2

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  true
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    performance/noAccumulatingSpread
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDescendingSpecificity
    style/noNonNullAssertion
    style/useArrayLiterals
    style/useConst
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useShorthandFunctionType
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

noImplicitAnyLet

### Playground link

https://biomejs.dev/playground/?code=aQBuAHQAZQByAGYAYQBjAGUAIABGAG8AbwAgAHsAIABhADoAIABuAHUAbQBiAGUAcgAgAH0ACgBkAGUAYwBsAGEAcgBlACAAYwBvAG4AcwB0ACAAZgBvAG8AOgAgAEYAbwBvAAoAZABlAGMAbABhAHIAZQAgAGMAbwBuAHMAdAAgAG4AOgAgAG4AdQBtAGIAZQByAAoAZABlAGMAbABhAHIAZQAgAGYAdQBuAGMAdABpAG8AbgAgAGYAbgAgACgAbQBhAHkAYgBlAEYAbwBvADoAIABGAG8AbwAgAHwAIAB1AG4AZABlAGYAaQBuAGUAZAApADoAIAB2AG8AaQBkAAoACgBsAGUAdAAgAG0AYQB5AGIAZQBGAG8AbwAKAGkAZgAgACgAbgAgADwAIAA1ADAAKQAgAHsACgAgACAAbQBhAHkAYgBlAEYAbwBvACAAPQAgAGYAbwBvAAoAfQAKAAoAZgBuACgAbQBhAHkAYgBlAEYAbwBvACkACgA%3D

### Expected result

`noImplicitAnyLet` should not trigger, since the type of `maybeFoo` is `Foo | undefined`, not `any`.

This can be verified in the TypeScript playground:

<img width="345" height="133" alt="Image" src="https://github.com/user-attachments/assets/dbf77881-5d41-4390-8adf-61b2fe12b95e" />

https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgGIHt3IN7LgLmRAFcBbAI2mQF8AoAEwgQBs4oUF0QBnMZGTIQzoGTVu2ScefEIRIVoolmxQxiIBGGBd+IZAApScAJ6VhQzMgA+ydYxigI9AJSEAbumD1atZhD5GphDCtMAwBnoAPMgArAAMzji0yMiBZpYAvPyYtHS0MCCGJunozrRAA

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by LinusU** [comment]
Actually, the if-statement might have been a red herring. It's just that all cases reported in our codebase was that pattern 😅 

Here is another example:

```ts
interface Foo { a: number }
interface Bar { b: number }
declare const foo: Foo
declare const bar: Bar
declare const n: number
declare function fn (maybeFoo: Foo | undefined): void

let maybeFoo
maybeFoo = foo

let maybeBar
maybeBar = bar

let test = maybeFoo
test = maybeBar

fn(maybeFoo)
fn(maybeBar)
```

TypeScript correctly reports errors on line 15 and 18, so the typings are definitely not `any`. Still, Biome reports "This variable implicitly has the `any` type." on line 8 and 11.

<img width="460" height="373" alt="Image" src="https://github.com/user-attachments/assets/0b9e20d8-0505-428a-a9e7-ee1a524f06c4" />

https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgGIHt3IN7LgLmRAFcBbAI2mQF8AoUSWRFAITih2XMJIqroAmEBABt2KBOhABnMMhiZCGdLSGjxySTLnl2hNlFXCxUCVNlEeZSobUmUMYiARhgU+SGQAKUnACelMpKmMgAPshOQjCgEAIAlIQAbujAArS0IhByvgEQyrQ5gSEAvPKY6ZnZ-pQGBdUQBsiluoYZWciQFqWFeeWdct31tbQwID71ynEjYz0GU0A

<img width="585" height="671" alt="Image" src="https://github.com/user-attachments/assets/b9392b6c-9da0-4ec6-8569-7a39dc51bac5" />

https://biomejs.dev/playground/?code=aQBuAHQAZQByAGYAYQBjAGUAIABGAG8AbwAgAHsAIABhADoAIABuAHUAbQBiAGUAcgAgAH0ACgBpAG4AdABlAHIAZgBhAGMAZQAgAEIAYQByACAAewAgAGIAOgAgAG4AdQBtAGIAZQByACAAfQAKAGQAZQBjAGwAYQByAGUAIABjAG8AbgBzAHQAIABmAG8AbwA6ACAARgBvAG8ACgBkAGUAYwBsAGEAcgBlACAAYwBvAG4AcwB0ACAAYgBhAHIAOgAgAEIAYQByAAoAZABlAGMAbABhAHIAZQAgAGMAbwBuAHMAdAAgAG4AOgAgAG4AdQBtAGIAZQByAAoAZABlAGMAbABhAHIAZQAgAGYAdQBuAGMAdABpAG8AbgAgAGYAbgAgACgAbQBhAHkAYgBlAEYAbwBvADoAIABGAG8AbwAgAHwAIAB1AG4AZABlAGYAaQBuAGUAZAApADoAIAB2AG8AaQBkAAoACgBsAGUAdAAgAG0AYQB5AGIAZQBGAG8AbwAKAG0AYQB5AGIAZQBGAG8AbwAgAD0AIABmAG8AbwAKAAoAbABlAHQAIABtAGEAeQBiAGUAQgBhAHIACgBtAGEAeQBiAGUAQgBhAHIAIAA9ACAAYgBhAHIACgAKAGwAZQB0ACAAdABlAHMAdAAgAD0AIABtAGEAeQBiAGUARgBvAG8ACgB0AGUAcwB0ACAAPQAgAG0AYQB5AGIAZQBCAGEAcgAKAAoAZgBuACgAbQBhAHkAYgBlAEYAbwBvACkACgBmAG4AKABtAGEAeQBiAGUAQgBhAHIAKQAKAA%3D%3D

**Comment by ematipico** [comment]
Closing because it's out of scope for this rule. This rule uses static analysis and doesn't leverage type inference.
